File: src/main/java/dark/leech/text/plugin/vbook/VBookPluginService.java
package dark.leech.text.plugin.vbook;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import dark.leech.text.action.Log;
import dark.leech.text.enities.PluginEntity;
import dark.leech.text.plugin.PluginManager;
import dark.leech.text.plugin.util.PluginPersistence;
import dark.leech.text.plugin.vbook.exception.VBookPluginException;
import dark.leech.text.plugin.vbook.exception.VBookRegistryException;
import dark.leech.text.plugin.vbook.model.VBookExtensionEntity;
import dark.leech.text.plugin.vbook.model.VBookPluginEntity;
import dark.leech.text.util.FileUtils;

/**
 * Service for orchestrating vBook plugin download, conversion, and installation. Provides
 * high-level operations for working with vBook extensions. Thread-safe with atomic installation
 * support.
 */
public class VBookPluginService {

    private static final String PLUGINS_DIR = "tools/plugins";
    private static final String VBOOK_CACHE_DIR = "tools/plugins/vbook-cache";

    private final VBookRepositoryClient repositoryClient;
    private final PluginZipExtractor extractor;
    private final VBookToLeechTextConverter converter;
    private final PluginManager pluginManager;

    // Lock for atomic installation operations
    private static final Lock installationLock = new ReentrantLock();

    // Track in-progress installations to prevent duplicates
    private static final Set<String> installingPlugins = new HashSet<>();

    public VBookPluginService() {
        this.repositoryClient = new VBookRepositoryClient();
        this.extractor = new PluginZipExtractor();
        this.converter = new VBookToLeechTextConverter();
        this.pluginManager = PluginManager.getManager();
    }

    /**
     * Get list of available plugins from a repository.
     *
     * @param repositoryUrl URL to plugin.json registry
     * @return List of available extensions
     * @throws VBookRegistryException if fetch fails
     */
    public List<VBookExtensionEntity> getAvailablePlugins(String repositoryUrl)
            throws VBookRegistryException {
        return repositoryClient.fetchExtensions(repositoryUrl);
    }

    /**
     * Get list of available plugins from default vBook repository.
     *
     * @return List of available extensions
     * @throws VBookRegistryException if fetch fails
     */
    public List<VBookExtensionEntity> getAvailablePlugins() throws VBookRegistryException {
        return getAvailablePlugins(VBookRepositoryClient.getDefaultRegistryUrl());
    }

    /**
     * Download and convert a vBook plugin from URL.
     *
     * @param pluginZipUrl URL to plugin.zip file
     * @return Converted PluginEntity
     * @throws VBookPluginException if download or conversion fails
     */
    public PluginEntity downloadAndConvert(String pluginZipUrl) throws VBookPluginException {
        try {
            // Extract from URL
            VBookPluginEntity vbookPlugin =
                    extractor.extractFromZip(new java.net.URL(pluginZipUrl));

            // Convert to LeechText format
            return converter.convert(vbookPlugin);

        } catch (Exception e) {
            throw new VBookPluginException("Failed to download plugin from: " + pluginZipUrl, e);
        }
    }

    /**
     * Install a vBook extension from repository with atomic operation.
     *
     * @param extension The extension to install
     * @return true if installation succeeded
     * @throws VBookPluginException if installation fails
     */
    public boolean installPlugin(VBookExtensionEntity extension) throws VBookPluginException {
        String pluginKey = extension.getName() + ":" + extension.getVersion();

        installationLock.lock();
        try {
            // Check if already installing (prevent duplicate installations)
            if (installingPlugins.contains(pluginKey)) {
                Log.add("Plugin installation already in progress: " + extension.getName());
                return false;
            }

            installingPlugins.add(pluginKey);

            try {
                // Download and convert
                PluginEntity pluginEntity = downloadAndConvert(extension.getPath());

                // Validate plugin before installation
                if (!PluginPersistence.validate(pluginEntity)) {
                    throw new VBookPluginException(
                            "Plugin validation failed: " + extension.getName());
                }

                // Check for name conflicts
                if (hasNameConflict(pluginEntity.getName())) {
                    pluginEntity.setName(pluginEntity.getName() + " (vBook)");
                }

                // Atomic save to plugins directory (uses temp file + atomic move)
                PluginPersistence.saveAtomic(pluginEntity);

                // Add to PluginManager
                pluginManager.add(pluginEntity);

                Log.add("Successfully installed plugin: " + pluginEntity.getName());
                return true;

            } finally {
                installingPlugins.remove(pluginKey);
            }

        } catch (Exception e) {
            Log.add("Failed to install plugin: " + extension.getName() + " - " + e.getMessage());
            throw new VBookPluginException("Failed to install plugin: " + extension.getName(), e);
        } finally {
            installationLock.unlock();
        }
    }

    /**
     * Find and install plugin by matching URL against available extensions. Errors are logged for
     * observability (not silent).
     *
     * @param targetUrl URL to match against plugin regex patterns
     * @return Installed PluginEntity, or null if no match found
     */
    public PluginEntity findAndInstallByUrl(String targetUrl) {
        if (targetUrl == null || targetUrl.isEmpty()) {
            Log.add("Auto-discovery: empty URL provided");
            return null;
        }

        try {
            // Get available plugins from default repository
            List<VBookExtensionEntity> extensions = getAvailablePlugins();

            // Find matching extension
            for (VBookExtensionEntity extension : extensions) {
                try {
                    // Download plugin to get its regex
                    VBookPluginEntity vbookPlugin =
                            extractor.extractFromZip(new java.net.URL(extension.getPath()));

                    // Check if URL matches plugin's regex
                    if (vbookPlugin.getRegexp() != null
                            && targetUrl.matches(vbookPlugin.getRegexp())) {
                        // Convert and install
                        PluginEntity pluginEntity = converter.convert(vbookPlugin);

                        if (hasNameConflict(pluginEntity.getName())) {
                            pluginEntity.setName(pluginEntity.getName() + " (vBook)");
                        }

                        // Use atomic installation
                        Path pluginPath = PluginPersistence.saveAtomic(pluginEntity);
                        pluginManager.add(pluginEntity);

                        Log.add(
                                "Auto-discovery: installed plugin "
                                        + pluginEntity.getName()
                                        + " for URL: "
                                        + targetUrl);
                        return pluginEntity;
                    }
                } catch (Exception e) {
                    // Log error but continue trying other plugins
                    Log.add(
                            "Auto-discovery: failed to process plugin "
                                    + extension.getName()
                                    + " - "
                                    + e.getMessage());
                }
            }

            Log.add("Auto-discovery: no matching plugin found for URL: " + targetUrl);
            return null;

        } catch (Exception e) {
            // Log error with context
            Log.add(
                    "Auto-discovery: error processing repository for URL: "
                            + targetUrl
                            + " - "
                            + e.getMessage());
            return null;
        }
    }

    /** Check if a plugin with the given name already exists. */
    private boolean hasNameConflict(String name) {
        if (name == null) return false;

        for (PluginEntity existing : pluginManager.list()) {
            if (name.equals(existing.getName())) {
                return true;
            }
        }
        return false;
    }

    /** Clear cached plugin files. */
    public void clearCache() {
        try {
            Path cacheDir = Paths.get(FileUtils.validate(VBOOK_CACHE_DIR));
            if (Files.exists(cacheDir)) {
                Files.walk(cacheDir)
                        .filter(p -> !p.equals(cacheDir))
                        .forEach(
                                p -> {
                                    try {
                                        Files.deleteIfExists(p);
                                    } catch (IOException e) {
                                        // Ignore
                                    }
                                });
            }
        } catch (IOException e) {
            // Ignore
        }
    }
}

File: engine/src/commonMain/kotlin/dev/haipham22/leechtext/plugin/vbook/VBookPluginService.kt
package dev.haipham22.leechtext.plugin.vbook

import dev.haipham22.leechtext.entities.PluginEntity
import dev.haipham22.leechtext.log.EngineLogger
import dev.haipham22.leechtext.plugin.PluginManager
import dev.haipham22.leechtext.plugin.util.PluginPersistence
import dev.haipham22.leechtext.plugin.vbook.exception.VBookPluginException
import dev.haipham22.leechtext.plugin.vbook.exception.VBookRegistryException
import dev.haipham22.leechtext.plugin.vbook.model.VBookExtensionEntity
import dev.haipham22.leechtext.util.EnginePaths
import dev.haipham22.leechtext.util.LeechJson
import dev.haipham22.leechtext.util.monitorLock
import dev.haipham22.leechtext.util.parseHost
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import okio.FileSystem

/**
 * Orchestrator download + convert + install vBook plugin (port từ
 * vbook/VBookPluginService.java). Thread-safe, atomic installation.
 */
class VBookPluginService(
    private val log: EngineLogger,
    private val repositoryClient: VBookRepositoryClient = VBookRepositoryClient(log),
    private val extractor: PluginZipExtractor = PluginZipExtractor(log),
    private val converter: VBookToLeechTextConverter = VBookToLeechTextConverter(),
    private val pluginPersistence: PluginPersistence = PluginPersistence(log),
    private val pluginManager: PluginManager,
) {
    companion object {
        // Lock cho atomic installation + track đang chạy — chống duplicate
        private val installationLock = Any()
        private val installingPlugins = HashSet<String>()
    }

    /** List plugin có sẵn từ repository. */
    @Throws(VBookRegistryException::class)
    fun getAvailablePlugins(repositoryUrl: String): List<VBookExtensionEntity> = repositoryClient.fetchExtensions(repositoryUrl)

    /** List plugin từ default vBook repository. */
    @Throws(VBookRegistryException::class)
    fun getAvailablePlugins(): List<VBookExtensionEntity> = getAvailablePlugins(VBookRepositoryClient.getDefaultRegistryUrl())

    private val fs: FileSystem get() = FileSystem.SYSTEM
    private val cacheFile: okio.Path get() = EnginePaths.dataDir / "tools" / "extensions-cache.json"

    /** Cache disk extension list — UI đổ vào state trước khi fetch (stale-while-revalidate). */
    fun loadCachedExtensions(): List<VBookExtensionEntity> = try {
        if (!fs.exists(cacheFile)) {
            emptyList()
        } else {
            runCatching { LeechJson.decodeFromString<List<VBookExtensionEntity>>(fs.read(cacheFile) { readUtf8() }) }.getOrNull() ?: emptyList()
        }
    } catch (e: Exception) {
        emptyList()
    }

    /** Ghi cache disk extension list sau fetch thành công. */
    fun saveCachedExtensions(exts: List<VBookExtensionEntity>) {
        try {
            cacheFile.parent?.let { fs.createDirectories(it) }
            fs.write(cacheFile) { writeUtf8(LeechJson.encodeToString(exts)) }
        } catch (_: Exception) {
            // Cache disk chỉ là tối ưu (stale-while-revalidate) — ghi fail không làm fail caller
        }
    }

    /** Download + convert vBook plugin từ URL. */
    @Throws(VBookPluginException::class)
    fun downloadAndConvert(pluginZipUrl: String): PluginEntity = try {
        val vbookPlugin = extractor.extractFromZip(pluginZipUrl)
        converter.convert(vbookPlugin)
    } catch (e: Exception) {
        throw VBookPluginException("Failed to download plugin from: $pluginZipUrl", e)
    }

    /**
     * Install vBook extension từ repository — atomic operation.
     *
     * @return true nếu install thành công
     */
    @Throws(VBookPluginException::class)
    fun installPlugin(extension: VBookExtensionEntity): Boolean {
        val pluginKey = "${extension.name}:${extension.version}"

        // Đảm bảo working directory tồn tại
        ensureWorkingDirectory()

        return monitorLock(installationLock) {
            // Chống duplicate installation
            if (pluginKey in installingPlugins) {
                log.add("Plugin installation already in progress: ${extension.name}")
                return@monitorLock false
            }

            installingPlugins.add(pluginKey)

            try {
                // Download + convert
                val pluginEntity = downloadAndConvert(extension.path ?: "")

                // Validate trước khi install
                if (!pluginPersistence.validate(pluginEntity)) {
                    throw VBookPluginException("Plugin validation failed: ${extension.name}")
                }

                // Check name conflict
                if (hasNameConflict(pluginEntity.name)) {
                    pluginEntity.name = pluginEntity.name + " (vBook)"
                }

                // Atomic save vào plugins dir (temp file + atomic move)
                pluginPersistence.saveAtomic(pluginEntity)

                // Add vào PluginManager
                pluginManager.add(pluginEntity)

                log.add("Successfully installed plugin: ${pluginEntity.name}")
                true
            } catch (e: Exception) {
                log.add("Failed to install plugin: ${extension.name} - ${e.message}")
                // Log full exception — add(String) chỉ in message, cause chain mất → không debug được
                log.add(e)
                throw VBookPluginException("Failed to install plugin: ${extension.name}", e)
            } finally {
                installingPlugins.remove(pluginKey)
            }
        }
    }

    /**
     * Tìm + install plugin match URL với extensions có sẵn. Error được log (không silent).
     *
     * Pre-filter theo host của extension.source — chỉ tải zip ứng viên khớp host, không
     * quét cả registry (17+ zip × repo = chờ vô ích). Hủy được giữa các zip qua
     * cancellation của coroutine gọi.
     *
     * @return PluginEntity đã install, null nếu không match
     */
    suspend fun findAndInstallByUrl(targetUrl: String?): PluginEntity? { // NOSONAR — cognitive 16/15: chuỗi try auto-discovery, tách hàm làm mẻ flow
        if (targetUrl.isNullOrEmpty()) {
            log.add("Auto-discovery: empty URL provided")
            return null
        }

        return try {
            val targetHost = parseHost(targetUrl)
            val extensions = getAvailablePlugins()

            val candidates =
                if (targetHost != null) {
                    val byHost =
                        extensions.filter { ext ->
                            val host = parseHost(ext.source.orEmpty())
                            host != null && (
                                host == targetHost ||
                                    host.removePrefix("www.") == targetHost.removePrefix("www.")
                                )
                        }
                    if (byHost.isEmpty()) {
                        log.add("Auto-discovery: no plugin source matches host $targetHost")
                        return null
                    }
                    byHost
                } else {
                    extensions
                }

            candidates.forEachIndexed { index, extension ->
                currentCoroutineContext().ensureActive()
                log.add("Auto-discovery: checking ${extension.name} (${index + 1}/${candidates.size})")
                try {
                    // Download plugin để lấy regex
                    val vbookPlugin = extractor.extractFromZip(extension.path.orEmpty())

                    // Check URL match regex — prefix scheme optional như PluginManager.get
                    // (không có thì URL https:// không bao giờ full-match)
                    if (vbookPlugin.regexp != null &&
                        targetUrl.matches(Regex("(https?://)?${vbookPlugin.regexp!!}"))
                    ) {
                        // Convert + install
                        val pluginEntity = converter.convert(vbookPlugin)

                        if (hasNameConflict(pluginEntity.name)) {
                            pluginEntity.name = pluginEntity.name + " (vBook)"
                        }

                        // Atomic installation
                        pluginPersistence.saveAtomic(pluginEntity)
                        pluginManager.add(pluginEntity)

                        log.add(
                            "Auto-discovery: installed plugin ${pluginEntity.name}" +
                                " for URL: $targetUrl",
                        )
                        return pluginEntity
                    }
                } catch (e: Exception) {
                    // Log nhưng thử plugin khác
                    log.add(
                        "Auto-discovery: failed to process plugin ${extension.name}" +
                            " - ${e.message}",
                    )
                }
            }

            log.add("Auto-discovery: no matching plugin found for URL: $targetUrl")
            null
        } catch (e: Exception) {
            log.add(
                "Auto-discovery: error processing repository for URL: $targetUrl" +
                    " - ${e.message}",
            )
            null
        }
    }

    /** Plugin với name này đã tồn tại chưa. */
    private fun hasNameConflict(name: String?): Boolean {
        if (name == null) return false
        return pluginManager.list().any { name == it.name }
    }

    /**
     * Đảm bảo working directory tồn tại + writable.
     *
     * @throws VBookPluginException nếu không tạo/ghi được
     */
    @Throws(VBookPluginException::class)
    private fun ensureWorkingDirectory() {
        try {
            val workingDir = EnginePaths.dataDir / "tools" / "plugins"
            fs.createDirectories(workingDir)
        } catch (e: Exception) {
            throw VBookPluginException("Failed to ensure working directory exists", e)
        }
    }

    /** Clear cache plugin files. */
    fun clearCache() {
        try {
            val cacheDir = EnginePaths.dataDir / "tools" / "plugins" / "vbook-cache"
            if (fs.exists(cacheDir)) {
                // Xóa file trước, dir sau (list không đệ quy — cache phẳng 1 cấp zip files)
                fs.list(cacheDir).forEach { file ->
                    runCatching {
                        if (fs.metadataOrNull(file)?.isDirectory == true) {
                            fs.deleteRecursively(file)
                        } else {
                            fs.delete(file)
                        }
                    }
                }
            }
        } catch (_: Exception) {
            // Clear cache không bao giờ ném — caller chỉ dọn dẹp, lỗi không có gì để xử lý
        }
    }
}
